docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+}
+
+#endregion
+
+#region Flow State Nullability
+
+public interface IArrayTypeSymbol : ITypeSymbol
+{
+    Nullability ElementNullability { get; }
+}
+
+public interface INamedTypeSymbol : ITypeSymbol
+{
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+}
+
+public interface ITypeParameterSymbol : ITypeSymbol
+{
+    Nullability ReferenceTypeConstraintNullability { get; }
+    ImmutableArray<Nullability> ConstraintsNullabilities { get; }
+}
+
+public struct TypeInfo
+{
+    public Nullability Nullability { get; }
+    public Nullability ConvertedNullability { get; }
+}
+
+#endregion
+
+#region Compilation
+
+public class Compilation
+{
+    public SemanticModel GetSemanticModel(bool skipNullabilityInformation);
+}
+
+
+#endregion
+
+#region Printing
+
+public interface ITypeSymbol
+{
+    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
+
+/*
+ * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
+ */
+
+#endregion
+```
+
+## Considered Alternatives
+
+### `Nullability` on `ITypeSymbol`
+
+The basic principle of this alternative proposal is to add `Nullability Nullability { get; }` to `ITypeSymbol`, and remove the rest of the proposed nullability retrieval APIs in the main section.
+This has a few benefits:
+
+1. Consumers of `ITypeSymbol`s won't have to keep track of where the type symbol came from to retrieve nullability information, they just have to ask the type.
+2. Printing APIs and code generation APIs do not need to have new overloads introduced to allow communication of top-level nullability information.
+3. Consumers of the new API don't have to separately worry about nullability when comparing types, as `string`, `string~`, and `string?` are truly different types.
+
+However, this approach has a few major issues that ultimately break our ability to do this design.
+
+1. This approach will violate principle #1 of our goals for this API, breaking existing analyzers and refactorings extensively.
+A common strategy for these APIs is to call `Compilation.GetTypeByMetadataName`, or one of the similar overloads, and then look for fields, locals, properties, or other similar symbols of the requested type.
+Up to this point, our guidance has been that, for non-generic types, it is perfectly fine to just compare these symbols via `==` if you're looking for exact matches, and that you only need to resort to `OriginalDefinition` for types with generics.
```

Comment:
i don't believe that's ever been the guidance.  For example, evne in Roslyn itself we continuously tell people that they need ot use .Equals even on non-generics since there is no guarantee of any sort of reference equality.

That said, even if users use .Equals i would *still* expect these types not to be the same.  `int?` is not the same type as `int` regardless of whether i use `==` or `.Equals`.  Similarly, `dynamic` is not the same as `object` and `(int, int)` is not the same as `(int a, int b)`.  These are different types, and my expectation as an analyzer author is that i should not consider the same.